// @shapeshiftoss/hdwallet-keepkey/dist/cosmos.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cosmosSignTx = exports.cosmosGetAddress = exports.cosmosGetAccountPaths = void 0;
const Messages = __importStar(require("@keepkey/device-protocol/lib/messages_pb"));
const CosmosMessages = __importStar(require("@keepkey/device-protocol/lib/messages-cosmos_pb"));
const core = __importStar(require("@shapeshiftoss/hdwallet-core"));
const bs58check_1 = __importDefault(require("bs58check"));
const p_lazy_1 = __importDefault(require("p-lazy"));
const protoTxBuilder = p_lazy_1.default.from(() => Promise.resolve().then(() => __importStar(require("@shapeshiftoss/proto-tx-builder"))));
function cosmosGetAccountPaths(msg) {
    return [
        {
            addressNList: [0x80000000 + 44, 0x80000000 + core.slip44ByCoin("Atom"), 0x80000000 + msg.accountIdx, 0, 0],
        },
    ];
}
exports.cosmosGetAccountPaths = cosmosGetAccountPaths;
function cosmosGetAddress(transport, msg) {
    return __awaiter(this, void 0, void 0, function* () {
        const getAddr = new CosmosMessages.CosmosGetAddress();
        getAddr.setAddressNList(msg.addressNList);
        getAddr.setShowDisplay(msg.showDisplay !== false);
        const response = yield transport.call(Messages.MessageType.MESSAGETYPE_COSMOSGETADDRESS, getAddr, {
            msgTimeout: core.LONG_TIMEOUT,
        });
        const cosmosAddress = response.proto;
        return core.mustBeDefined(cosmosAddress.getAddress());
    });
}
exports.cosmosGetAddress = cosmosGetAddress;
function cosmosSignTx(transport, msg) {
    return __awaiter(this, void 0, void 0, function* () {
        const address = yield cosmosGetAddress(transport, { addressNList: msg.addressNList });
        const getPublicKeyMsg = new Messages.GetPublicKey();
        getPublicKeyMsg.setAddressNList(msg.addressNList);
        getPublicKeyMsg.setEcdsaCurveName("secp256k1");
        const pubkeyMsg = (yield transport.call(Messages.MessageType.MESSAGETYPE_GETPUBLICKEY, getPublicKeyMsg, {
            msgTimeout: core.DEFAULT_TIMEOUT,
        })).proto;
        const pubkey = bs58check_1.default.decode(core.mustBeDefined(pubkeyMsg.getXpub())).slice(45);
        return transport.lockDuring(() => __awaiter(this, void 0, void 0, function* () {
            const signTx = new CosmosMessages.CosmosSignTx();
            signTx.setAddressNList(msg.addressNList);
            signTx.setAccountNumber(msg.account_number);
            signTx.setChainId(msg.chain_id);
            signTx.setFeeAmount(parseInt(msg.tx.fee.amount[0].amount));
            signTx.setGas(parseInt(msg.tx.fee.gas));
            signTx.setSequence(msg.sequence);
            if (msg.tx.memo !== undefined) {
                signTx.setMemo(msg.tx.memo);
            }
            signTx.setMsgCount(1);
            let resp = yield transport.call(Messages.MessageType.MESSAGETYPE_COSMOSSIGNTX, signTx, {
                msgTimeout: core.LONG_TIMEOUT,
                omitLock: true,
            });
            for (const m of msg.tx.msg) {
                if (resp.message_enum !== Messages.MessageType.MESSAGETYPE_COSMOSMSGREQUEST) {
                    throw new Error(`cosmos: unexpected response ${resp.message_type}`);
                }
                let ack;
                if (m.type === "cosmos-sdk/MsgSend") {
                    if (m.value.amount.length !== 1) {
                        throw new Error("cosmos: Multiple amounts per msg not supported");
                    }
                    const denom = m.value.amount[0].denom;
                    if (denom !== "uatom") {
                        throw new Error("cosmos: Unsupported denomination: " + denom);
                    }
                    const send = new CosmosMessages.CosmosMsgSend();
                    send.setFromAddress(m.value.from_address);
                    send.setToAddress(m.value.to_address);
                    send.setAmount(m.value.amount[0].amount);
                    ack = new CosmosMessages.CosmosMsgAck();
                    ack.setSend(send);
                }
                else if (m.type === "cosmos-sdk/MsgDelegate") {
                    const denom = m.value.amount.denom;
                    if (denom !== "uatom") {
                        throw new Error("cosmos: Unsupported denomination: " + denom);
                    }
                    const delegate = new CosmosMessages.CosmosMsgDelegate();
                    delegate.setDelegatorAddress(m.value.delegator_address);
                    delegate.setValidatorAddress(m.value.validator_address);
                    delegate.setAmount(m.value.amount.amount);
                    ack = new CosmosMessages.CosmosMsgAck();
                    ack.setDelegate(delegate);
                }
                else if (m.type === "cosmos-sdk/MsgUndelegate") {
                    const denom = m.value.amount.denom;
                    if (denom !== "uatom") {
                        throw new Error("cosmos: Unsupported denomination: " + denom);
                    }
                    const undelegate = new CosmosMessages.CosmosMsgUndelegate();
                    undelegate.setDelegatorAddress(m.value.delegator_address);
                    undelegate.setValidatorAddress(m.value.validator_address);
                    undelegate.setAmount(m.value.amount.amount);
                    ack = new CosmosMessages.CosmosMsgAck();
                    ack.setUndelegate(undelegate);
                }
                else if (m.type === "cosmos-sdk/MsgBeginRedelegate") {
                    const denom = m.value.amount.denom;
                    if (denom !== "uatom") {
                        throw new Error("cosmos: Unsupported denomination: " + denom);
                    }
                    const redelegate = new CosmosMessages.CosmosMsgRedelegate();
                    redelegate.setDelegatorAddress(m.value.delegator_address);
                    redelegate.setValidatorSrcAddress(m.value.validator_src_address);
                    redelegate.setValidatorDstAddress(m.value.validator_dst_address);
                    redelegate.setAmount(m.value.amount.amount);
                    ack = new CosmosMessages.CosmosMsgAck();
                    ack.setRedelegate(redelegate);
                }
                else if (m.type === "cosmos-sdk/MsgWithdrawDelegationReward") {
                    const rewards = new CosmosMessages.CosmosMsgRewards();
                    rewards.setDelegatorAddress(m.value.delegator_address);
                    rewards.setValidatorAddress(m.value.validator_address);
                    if (m.value.amount) {
                        const denom = m.value.amount.denom;
                        if (denom !== "uatom") {
                            throw new Error("cosmos: Unsupported denomination: " + denom);
                        }
                        rewards.setAmount(m.value.amount.amount);
                    }
                    ack = new CosmosMessages.CosmosMsgAck();
                    ack.setRewards(rewards);
                }
                else if (m.type === "cosmos-sdk/MsgTransfer") {
                    const denom = m.value.token.denom;
                    if (denom !== "uatom") {
                        throw new Error("cosmos: Unsupported denomination: " + denom);
                    }
                    const ibcTransfer = new CosmosMessages.CosmosMsgIBCTransfer();
                    ibcTransfer.setReceiver(m.value.receiver);
                    ibcTransfer.setSender(m.value.sender);
                    ibcTransfer.setSourceChannel(m.value.source_channel);
                    ibcTransfer.setSourcePort(m.value.source_port);
                    ibcTransfer.setRevisionHeight(m.value.timeout_height.revision_height);
                    ibcTransfer.setRevisionNumber(m.value.timeout_height.revision_number);
                    ibcTransfer.setAmount(m.value.token.amount);
                    ibcTransfer.setDenom(m.value.token.denom);
                    ack = new CosmosMessages.CosmosMsgAck();
                    ack.setIbcTransfer(ibcTransfer);
                }
                else {
                    throw new Error(`cosmos: Message ${m.type} is not yet supported`);
                }
                resp = yield transport.call(Messages.MessageType.MESSAGETYPE_COSMOSMSGACK, ack, {
                    msgTimeout: core.LONG_TIMEOUT,
                    omitLock: true,
                });
            }
            if (resp.message_enum !== Messages.MessageType.MESSAGETYPE_COSMOSSIGNEDTX) {
                throw new Error(`cosmos: unexpected response ${resp.message_type}`);
            }
            const signedTx = resp.proto;
            const offlineSigner = {
                getAccounts() {
                    return __awaiter(this, void 0, void 0, function* () {
                        return [
                            {
                                address,
                                algo: "secp256k1",
                                pubkey,
                            },
                        ];
                    });
                },
                signAmino(signerAddress, signDoc) {
                    return __awaiter(this, void 0, void 0, function* () {
                        if (signerAddress !== address)
                            throw new Error("expected signerAddress to match address");
                        return {
                            signed: signDoc,
                            signature: {
                                pub_key: {
                                    type: "tendermint/PubKeySecp256k1",
                                    value: signedTx.getPublicKey_asB64(),
                                },
                                signature: signedTx.getSignature_asB64(),
                            },
                        };
                    });
                },
            };
            const signerData = {
                sequence: Number(msg.sequence),
                accountNumber: Number(msg.account_number),
                chainId: msg.chain_id,
            };
            return (yield protoTxBuilder).sign(address, msg.tx, offlineSigner, signerData);
        }));
    });
}
exports.cosmosSignTx = cosmosSignTx;
